dm-agent/src/main.rs

```diff
@@ -0,0 +1,354 @@
+
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    #[error(transparent)]
+    MqttError(#[from] Arc<mqtt_client::Error>), //TODO: avoid the logical duplication of this error and MqttClientError because of the Arc
+
+    #[error(transparent)]
+    MqttClientError(#[from] mqtt_client::Error),
+
+    #[error(
+        "Message received on invalid collectd topic: {0}. \
+        Collectd message topics must be in the format collectd/<hostname>/<metric-plugin-name>/<metric-key>"
+    )]
+    InvalidMeasurementTopic(String),
+
+    #[error("Non UTF-8 payload received on topic: {0}")]
+    InvalidMeasurementPayload(String),
+
+    #[error(
+        "Invalid payload: {0} received on topic: {1}. Expected payload format: <timestamp>:<value>"
+    )]
+    InvalidMeasurementPayloadFormat(String, String),
+
+    #[error("Invalid measurement value: {0} received on topic: {1}. Must be a number")]
+    InvalidMeasurementValue(String, String),
+
+    #[error("Invalid Thin Edge JSON measurement format")]
+    InvalidThinEdgeJsonError(#[from] thin_edge_json::group::MeasurementGrouperError),
+
+    #[error(transparent)]
+    ThinEdgeJsonSerializationError(#[from] ThinEdgeJsonSerializationError),
+}
+
+struct CollectdMessage {
+    metric_group_key: String,
+    metric_key: String,
+    metric_value: f64,
+}
+
+impl TryFrom<Message> for CollectdMessage {
+    type Error = Error;
+
+    fn try_from(mqtt_message: Message) -> Result<Self, Error> {
+        let topic = mqtt_message.topic.name.clone();
+        let split = topic.split('/').collect::<Vec<&str>>();
+        if split.len() == 4 {
+            let metric_group_key = split[METRIC_GROUP_KEY_POS].into();
+            let metric_key = split[METRIC_KEY_POS].into();
+
+            let payload = mqtt_message.payload;
+            let payload = from_utf8(&payload)
+                .map_err(|_err| Error::InvalidMeasurementPayload(topic.clone()))?; //TODO: avoid topic clone?
+            let split = payload.split(':').collect::<Vec<&str>>();
+
+            if split.len() == 2 {
+                let metric_value = split[1]
+                    .trim_end_matches(char::from(0))
+                    .parse::<f64>()
+                    .map_err(|_err| {
+                        Error::InvalidMeasurementValue(split[1].into(), topic.clone())
+                    })?;
+
+                Ok(Self {
+                    metric_group_key,
+                    metric_key,
+                    metric_value,
+                })
+            } else {
+                return Err(Error::InvalidMeasurementPayloadFormat(
+                    payload.into(),
+                    topic.clone(),
+                ));
+            }
+        } else {
+            //TODO: Why explicit return
+            return Err(Error::InvalidMeasurementTopic(topic));
+        }
+    }
+}
+
+#[derive(Debug)]
+struct ThinEdgeMessage {
+    message_grouper: MeasurementGrouper,
+}
+
+impl ThinEdgeMessage {
```

Comment:
I have a mix-feeling with this struct which responsibility is not clear.

* A facade over the measurement grouper?
* A manager of the measurement collection (as suggested by the fact this struct assigns a timestamp)?
* A manager of the output over MQTT ?

Reply:
This struct was introduced to cover the first two points. The third one could be moved out.